src/encoder.rs

```diff
@@ -99,29 +111,159 @@ impl From<FormatErrorKind> for FormatError {
+        let green_x = self.green.0.into_scaled().to_be_bytes();
+        let green_y = self.green.1.into_scaled().to_be_bytes();
+        let blue_x  = self.blue.0.into_scaled().to_be_bytes();
+        let blue_y  = self.blue.1.into_scaled().to_be_bytes();
+        [
+            white_x[0], white_x[1], white_x[2], white_x[3],
+            white_y[0], white_y[1], white_y[2], white_y[3],
+            red_x[0],   red_x[1],   red_x[2],   red_x[3],
+            red_y[0],   red_y[1],   red_y[2],   red_y[3],
+            green_x[0], green_x[1], green_x[2], green_x[3],
+            green_y[0], green_y[1], green_y[2], green_y[3],
+            blue_x[0],  blue_x[1],  blue_x[2],  blue_x[3],
+            blue_y[0],  blue_y[1],  blue_y[2],  blue_y[3],
+        ]
+    }
+
+    pub fn encode<W: Write>(self, w: &mut W) -> io::Result<()> {
+        chunk::encode_chunk(w, chunk::cHRM, &self.to_be_bytes())
+    }
+}
+
+impl SrgbRenderingIntent {
+    pub fn encode<W: Write>(self, w: &mut W) -> io::Result<()> {
+        chunk::encode_chunk(w, chunk::sRGB, &[self.into_raw()])
+    }
+}
+
+impl Info {
+    pub fn encode<W: Write>(&self, w: &mut W) -> io::Result<()> {
+        chunk::IHDR_encode(
+            w,
+            self.width,
+            self.height,
+            self.bit_depth,
+            self.color_type,
+            self.interlaced,
+        )?;
+
+        if let Some(p) = &self.palette {
+            chunk::encode_chunk(w, chunk::PLTE, p)?;
+        };
+
+        if let Some(t) = &self.trns {
+            chunk::encode_chunk(w, chunk::tRNS, t)?;
+        }
+
+        // If specified, the sRGB information overrides the source gamma and chromaticities.
+        if let Some(srgb) = &self.srgb {
+            let gamma = crate::srgb::substitute_gamma();
+            let chromaticities = crate::srgb::substitute_chromaticities();
+            srgb.encode(w)?;
+            gamma.encode(w)?;
+            chromaticities.encode(w)?;
+        } else {
+            self.source_gamma.map_or(Ok(()), |v| v.encode(w))?;
+            self.source_chromaticities.map_or(Ok(()), |v| v.encode(w))?;
+        }
+        self.animation_control.map_or(Ok(()), |v| v.encode(w))?;
+
+        Ok(())
+    }
+}
+
+/// PNG Animation Encoder
+pub struct Encoder<W: io::Write> {
     w: W,
     info: Info,
+    sep_def_img: bool,
+    adaptive: AdaptiveFilterType,
     filter: FilterType,
-    adaptive_filter: AdaptiveFilterType,
 }
 
 impl<W: Write> Encoder<W> {
-    pub fn new(w: W, width: u32, height: u32) -> Encoder<W> {
+    pub fn new(w: W, width: u32, height: u32) -> Self {
         let mut info = Info::default();
         info.width = width;
         info.height = height;
-        Encoder {
+        Self {
             w,
             info,
-            filter: FilterType::default(),
-            adaptive_filter: AdaptiveFilterType::default(),
+            sep_def_img: false,
+            filter: <_>::default(),
+            adaptive: <_>::default(),
```

Comment:
The type annotations make this a lot more readable, if one wants to understand what is happening. Another unecessary clever change. You seem to want to try out stuff but changing working, existing code is not the proper outlet for this.